modules/mongodb/scanner.go

```diff
@@ -0,0 +1,346 @@
+	return new(Scanner)
+}
+
+// StartScan opens a connection to the target and sets up a scan instance for it.
+func (scanner *Scanner) StartScan(target *zgrab2.ScanTarget) (*scan, error) {
+	conn, err := target.Open(&scanner.config.BaseFlags)
+	if err != nil {
+		return nil, err
+	}
+
+	return &scan{
+		target:  target,
+		scanner: scanner,
+		result:  &Result{},
+		conn: &Connection{
+			scanner: scanner,
+			conn:    conn,
+		},
+		close: func() { conn.Close() },
+	}, nil
+}
+
+// IsMasterResult hold the result of an isMaster command. Currently,
+// only interested in maxWireVersion.
+type IsMasterResult struct {
+	MaxWireVersion int32 `bson:"maxWireVersion"`
+}
+
+// getMaxWireVersion retrieves the maxWireVersion value reported by the MongoDB server.
+func getMaxWireVersion(conn *Connection) (int32, error) {
+	document := &IsMasterResult{}
+	doc_offset := MSGHEADER_LEN + 20
+	conn.Write(conn.scanner.isMasterMsg)
+
+	msg, err := conn.ReadMsg()
+	if err != nil {
+		return 0, err
+	}
+
+	if len(msg) < MSGHEADER_LEN + 4 {
+		err = fmt.Errorf("Server truncated message - no query reply (%d bytes: %s)", len(msg), hex.EncodeToString(msg))
+		return 0, err
+	}
+	respFlags := binary.LittleEndian.Uint32(msg[MSGHEADER_LEN:MSGHEADER_LEN + 5])
+	if respFlags & QUERY_RESP_FAILED != 0 {
+		err = fmt.Errorf("isMaster query failed")
+		return 0, err
+	}
+	doclen := int(binary.LittleEndian.Uint32(msg[doc_offset:doc_offset + 4]))
+	if len(msg[doc_offset:]) < doclen {
+		err = fmt.Errorf("Server truncated BSON reply doc (%d bytes: %s)",
+			  len(msg[doc_offset:]), hex.EncodeToString(msg))
+		return 0, err
+	}
+	err = bson.Unmarshal(msg[doc_offset:], &document)
+	if err != nil {
+		err = fmt.Errorf("Server sent invalid BSON reply doc (%d bytes: %s)",
+			  len(msg[doc_offset:]), hex.EncodeToString(msg))
+		return 0, err
+	}
+	return document.MaxWireVersion, nil
+}
+
+// Scan connects to a host and performs a scan.
+func (scanner *Scanner) Scan(target zgrab2.ScanTarget) (zgrab2.ScanStatus, interface{}, error) {
+	scan, err := scanner.StartScan(&target)
+	if err != nil {
+		return zgrab2.TryGetScanStatus(err), nil, err
+	}
+	defer scan.Close()
+
+	result := scan.result
+	max_wirev, err := getMaxWireVersion(scan.conn)
+	if err != nil {
+		return zgrab2.SCAN_PROTOCOL_ERROR, nil, err
+	}
+
+	var query []byte
+	var resplen_offset int
+	var resp_offset int
+
+	// Gleaned from wireshark - if "MaxWireVersion" is less than 7, then
+	// "build info" command should be sent in an OP_COMMAND with the query sent
+	// and response retrieved at "metadata" offset. At 7 and above, should 
+	// be sent as an OP_MSG in the "section" field, and response is at "body" offset
+	if max_wirev < 7 {
+		query = scanner.buildInfoCommandMsg
+		resplen_offset = 4
+		resp_offset = 0
+	} else {
+		query = scanner.buildInfoOpMsg
+		resplen_offset = 5
+		resp_offset = 5
+	}
+
+	scan.conn.Write(query)
+	msg, err := scan.conn.ReadMsg()
+	if err != nil {
+		return zgrab2.TryGetScanStatus(err), nil, err
```

Comment:
Is the data returned from the first packet / the fact that `getMaxWireVersion()` didn't throw sufficient to say that this is a MongoDB server?

If so, returning the error is still correct, but instead of nil, it should return a partial response object.

If OTOH random data could get this far, then returning a nil result is the right choice.

Reply:
Yes, if getMaxWireVersion() succeeds, then the server is speaking valid mongodb protocol. Should return partial response.